File: cloudflare/shared/engine.mjs
import { clamp, clamp01 } from "./helpers.mjs";

export function computeEventQuality(event) {
  if (typeof event.data_quality_score === "number" && event.data_quality_score > 0) {
    return clamp01(event.data_quality_score);
  }

  const distinctBooks = new Set((event.markets || []).map((market) => market.bookmaker)).size;
  const marketCount = (event.markets || []).length;
  const sourceCount = Array.isArray(event.sources) && event.sources.length > 0 ? event.sources.length : 1;

  const bookScore = Math.min(1, distinctBooks / 6);
  const marketScore = Math.min(1, marketCount / 12);
  const sourceScore = Math.min(1, sourceCount / 2);

  return clamp01(0.55 * bookScore + 0.3 * sourceScore + 0.15 * marketScore);
}

function capForOdd(limit, odd) {
  const caps = [];
  if (typeof limit.max_stake === "number" && limit.max_stake > 0) {
    caps.push(limit.max_stake);
  }
  if (typeof limit.max_payout === "number" && limit.max_payout > 0 && odd > 0) {
    caps.push(limit.max_payout / odd);
  }
  if (caps.length === 0) {
    return null;
  }
  return Math.min(...caps);
}

function effectiveOdd(bookmaker, odd, bookmakerLimits) {
  const limit = bookmakerLimits?.[bookmaker];
  if (!limit) {
    return odd;
  }
  const commission = clamp(Number(limit.commission_pct || 0), 0, 100);
  return 1 + (odd - 1) * (1 - commission / 100);
}

function groupByMarket(markets) {
  const grouped = new Map();
  for (const market of markets || []) {
    const key = market.market_key;
    if (!grouped.has(key)) {
      grouped.set(key, []);
    }
    grouped.get(key).push(market);
  }
  return grouped;
}

function bestOutcomes(marketQuotes, bookmakerLimits) {
  const bestOdds = {};
  const bestBooks = {};
  const outcomesSeen = new Set();

  for (const quote of marketQuotes || []) {
    for (const outcome of quote.outcomes || []) {
      outcomesSeen.add(outcome.key);
      const eff = effectiveOdd(quote.bookmaker, Number(outcome.odds), bookmakerLimits);
      if (!(outcome.key in bestOdds) || eff > bestOdds[outcome.key]) {
        bestOdds[outcome.key] = eff;
        bestBooks[outcome.key] = quote.bookmaker;
      }
    }
  }

  if (outcomesSeen.size < 2) {
    return null;
  }
  if (Object.keys(bestOdds).length !== outcomesSeen.size) {
    return null;
  }
  return { bestOdds, bestBooks };
}

function buildStakePlan(odds, books, bankroll, bookmakerLimits) {
  const implied = {};
  let totalImplied = 0;

  for (const [outcomeKey, odd] of Object.entries(odds)) {
    const probability = 1 / odd;
    implied[outcomeKey] = probability;
    totalImplied += probability;
  }

  const allocations = {};
  for (const [outcomeKey, probability] of Object.entries(implied)) {
    allocations[outcomeKey] = bankroll * (probability / totalImplied);
  }

  let usedBankroll = bankroll;
  let constraintsApplied = false;
  const notes = [];

  const scales = [];
  for (const [outcomeKey, stake] of Object.entries(allocations)) {
    const bookmaker = books[outcomeKey] || "";
    const limit = bookmakerLimits?.[bookmaker];
    if (!limit) {
      continue;
    }
    const cap = capForOdd(limit, odds[outcomeKey]);
    if (typeof cap === "number" && stake > cap) {
      constraintsApplied = true;
      notes.push(`${bookmaker}:${outcomeKey} cap=${cap.toFixed(2)}`);
      if (stake > 0) {
        scales.push(cap / stake);
      }
    }
  }

  if (scales.length > 0) {
    const scale = clamp(Math.min(...scales), 0, 1);
    for (const key of Object.keys(allocations)) {
      allocations[key] *= scale;
    }
    usedBankroll = Object.values(allocations).reduce((sum, value) => sum + value, 0);
  }

  for (const [outcomeKey, stake] of Object.entries(allocations)) {
    const bookmaker = books[outcomeKey] || "";
    const limit = bookmakerLimits?.[bookmaker];
    if (!limit) {
      continue;
    }
    const minStake = Math.max(0, Number(limit.min_stake || 0));
    if (stake < minStake) {
      return null;
    }
  }

  let guaranteedReturn = Number.POSITIVE_INFINITY;
  for (const key of Object.keys(allocations)) {
    guaranteedReturn = Math.min(guaranteedReturn, allocations[key] * odds[key]);
  }
  const guaranteedProfit = guaranteedReturn - usedBankroll;
  const guaranteedProfitPct = usedBankroll > 0 ? (guaranteedProfit / usedBankroll) * 100 : 0;

  return {
    requested_bankroll: bankroll,
    bankroll: usedBankroll,
    total_implied_probability: totalImplied,
    allocations,
    guaranteed_return: guaranteedReturn,
    guaranteed_profit: guaranteedProfit,
    guaranteed_profit_pct: guaranteedProfitPct,
    constraints_applied: constraintsApplied,
    constraints_note: notes.join("; "),
  };
}

export function scanEvents(events, options) {
  const opportunities = [];

  const minMarginPct = Number(options.min_margin_pct || 0);
  const minDistinctBooks = Math.max(2, Number(options.min_distinct_books || 2));
  const minEventSources = Math.max(1, Number(options.min_event_sources || 1));
  const minDataQuality = clamp01(Number(options.min_data_quality || 0));
  const slippageBps = Math.max(0, Number(options.slippage_bps || 0));
  const rejectionRate = clamp(Number(options.rejection_rate || 0), 0, 1);
  const bankroll = Number(options.bankroll || 1000);
  const bookmakerLimits = options.bookmaker_limits || {};

  for (const event of events || []) {
    const sourceCount = Array.isArray(event.sources) && event.sources.length > 0 ? event.sources.length : 1;
    if (sourceCount < minEventSources) {
      continue;
    }

    const quality = computeEventQuality(event);
    if (quality < minDataQuality) {
      continue;
    }

    const grouped = groupByMarket(event.markets || []);
    for (const [marketKey, marketQuotes] of grouped.entries()) {
      const best = bestOutcomes(marketQuotes, bookmakerLimits);
      if (!best) {
        continue;
      }

      const { bestOdds, bestBooks } = best;
      const totalImplied = Object.values(bestOdds).reduce((sum, odd) => sum + 1 / odd, 0);
      if (totalImplied >= 1) {
        continue;
      }

      const marginPct = (1 / totalImplied - 1) * 100;
      if (marginPct < minMarginPct) {
        continue;
      }
      if (new Set(Object.values(bestBooks)).size < minDistinctBooks) {
        continue;
      }

      const stakePlan = buildStakePlan(bestOdds, bestBooks, bankroll, bookmakerLimits);
      if (!stakePlan) {
        continue;
      }

      const slippagePenalty = (stakePlan.bankroll * slippageBps) / 10000;
      const rejectionPenalty = stakePlan.guaranteed_profit * rejectionRate;
      const riskPenalty = slippagePenalty + rejectionPenalty;
      const expectedProfit = stakePlan.guaranteed_profit - riskPenalty;
      const riskScore = clamp(rejectionRate * 65 + slippageBps / 4 + (1 - quality) * 35, 0, 100);

      opportunities.push({
        event_id: event.event_id,
        event: `${event.home} vs ${event.away}`,
        market: marketKey,
        outcomes: bestOdds,
        bookmakers: bestBooks,
        implied: totalImplied,
        margin_pct: marginPct,
        requested_bankroll: stakePlan.requested_bankroll,
        bankroll: stakePlan.bankroll,
        stakes: stakePlan.allocations,
        guaranteed_return: stakePlan.guaranteed_return,
        profit: stakePlan.guaranteed_profit,
        profit_pct: stakePlan.guaranteed_profit_pct,
        expected_profit: expectedProfit,
        risk_penalty: riskPenalty,
        risk_score: riskScore,
        source_count: sourceCount,
        data_quality_score: quality,
        constraints_applied: stakePlan.constraints_applied,
        constraints_note: stakePlan.constraints_note,
      });
    }
  }

  opportunities.sort(
    (a, b) =>
      b.expected_profit - a.expected_profit ||
      b.margin_pct - a.margin_pct ||
      b.data_quality_score - a.data_quality_score,
  );

  return opportunities;
}

File: src/iddiaarb/engine.py
from __future__ import annotations

from collections import defaultdict
from typing import Dict, Iterable, Tuple

from iddiaarb.limits import BookmakerLimit
from iddiaarb.models import ArbitrageOpportunity, Event, MarketQuote, StakePlan
from iddiaarb.quality import compute_event_quality


class ArbitrageEngine:
    def __init__(
        self,
        min_margin_pct: float = 0.0,
        min_distinct_books: int = 2,
        min_event_sources: int = 1,
        min_data_quality: float = 0.0,
        bookmaker_limits: dict[str, BookmakerLimit] | None = None,
        slippage_bps: float = 0.0,
        rejection_rate: float = 0.0,
    ) -> None:
        self.min_margin_pct = min_margin_pct
        self.min_distinct_books = min_distinct_books
        self.min_event_sources = max(1, min_event_sources)
        self.min_data_quality = max(0.0, min(1.0, min_data_quality))
        self.bookmaker_limits = bookmaker_limits or {}
        self.slippage_bps = max(0.0, slippage_bps)
        self.rejection_rate = max(0.0, min(1.0, rejection_rate))

    def scan(self, events: Iterable[Event], bankroll: float) -> list[ArbitrageOpportunity]:
        opportunities: list[ArbitrageOpportunity] = []
        for event in events:
            source_count = len(event.sources) if event.sources else 1
            if source_count < self.min_event_sources:
                continue

            quality = compute_event_quality(event)
            if quality < self.min_data_quality:
                continue

            grouped = self._group_by_market(event.markets)
            for market_key, market_quotes in grouped.items():
                result = self._best_outcomes(market_quotes)
                if not result:
                    continue
                best_odds, best_books = result
                total_implied = sum(1 / odd for odd in best_odds.values())
                if total_implied >= 1:
                    continue

                margin_pct = (1 / total_implied - 1) * 100
                if margin_pct < self.min_margin_pct:
                    continue
                if len(set(best_books.values())) < self.min_distinct_books:
                    continue

                stake_plan = self._build_stake_plan(best_odds, best_books, bankroll)
                if stake_plan is None:
                    continue

                slippage_penalty = (stake_plan.bankroll * self.slippage_bps) / 10000.0
                rejection_penalty = stake_plan.guaranteed_profit * self.rejection_rate
                risk_penalty = slippage_penalty + rejection_penalty
                expected_profit = stake_plan.guaranteed_profit - risk_penalty
                risk_score = max(
                    0.0,
                    min(
                        100.0,
                        (self.rejection_rate * 65.0)
                        + (self.slippage_bps / 4.0)
                        + ((1.0 - quality) * 35.0),
                    ),
                )
                opportunities.append(
                    ArbitrageOpportunity(
                        event_id=event.event_id,
                        event_label=f"{event.home} vs {event.away}",
                        market_key=market_key,
                        outcomes=best_odds,
                        bookmakers=best_books,
                        total_implied_probability=total_implied,
                        margin_pct=margin_pct,
                        stake_plan=stake_plan,
                        expected_profit=expected_profit,
                        risk_penalty=risk_penalty,
                        risk_score=risk_score,
                        source_count=source_count,
                        data_quality_score=quality,
                    )
                )
        opportunities.sort(
            key=lambda x: (x.expected_profit, x.margin_pct, x.data_quality_score), reverse=True
        )
        return opportunities

    def _group_by_market(self, markets: Iterable[MarketQuote]) -> Dict[str, list[MarketQuote]]:
        grouped: Dict[str, list[MarketQuote]] = defaultdict(list)
        for market in markets:
            grouped[market.market_key].append(market)
        return grouped

    def _best_outcomes(
        self, market_quotes: Iterable[MarketQuote]
    ) -> Tuple[Dict[str, float], Dict[str, str]] | None:
        best_odds: Dict[str, float] = {}
        best_books: Dict[str, str] = {}
        outcomes_seen = set()

        for quote in market_quotes:
            for outcome in quote.outcomes:
                outcomes_seen.add(outcome.key)
                effective_odd = self._effective_odd(quote.bookmaker, outcome.odds)
                if outcome.key not in best_odds or effective_odd > best_odds[outcome.key]:
                    best_odds[outcome.key] = effective_odd
                    best_books[outcome.key] = quote.bookmaker

        # Need at least 2 outcomes to define a hedge
        if len(outcomes_seen) < 2:
            return None
        if len(best_odds) != len(outcomes_seen):
            return None
        return best_odds, best_books

    def _effective_odd(self, bookmaker: str, odd: float) -> float:
        limit = self.bookmaker_limits.get(bookmaker)
        if not limit:
            return odd
        commission = max(0.0, min(100.0, limit.commission_pct))
        return 1 + ((odd - 1) * (1 - (commission / 100.0)))

    def _build_stake_plan(
        self, odds: Dict[str, float], books: Dict[str, str], bankroll: float
    ) -> StakePlan | None:
        implied = {k: 1 / v for k, v in odds.items()}
        total_implied = sum(implied.values())
        allocations = {k: bankroll * (p / total_implied) for k, p in implied.items()}
        used_bankroll = bankroll
        constraints_applied = False
        notes: list[str] = []

        scale_caps: list[float] = []
        for outcome_key, stake in allocations.items():
            bookmaker = books.get(outcome_key, "")
            limit = self.bookmaker_limits.get(bookmaker)
            if not limit:
                continue
            cap = limit.cap_for_odd(odds[outcome_key])
            if cap is not None and stake > cap:
                constraints_applied = True
                notes.append(f"{bookmaker}:{outcome_key} cap={cap:.2f}")
                if stake > 0:
                    scale_caps.append(cap / stake)
        if scale_caps:
            scale = min(scale_caps)
            scale = max(0.0, min(1.0, scale))
            allocations = {k: v * scale for k, v in allocations.items()}
            used_bankroll = sum(allocations.values())

        for outcome_key, stake in allocations.items():
            bookmaker = books.get(outcome_key, "")
            limit = self.bookmaker_limits.get(bookmaker)
            if not limit:
                continue
            if stake < max(0.0, limit.min_stake):
                return None

        guaranteed_return = min(allocations[k] * odds[k] for k in allocations)
        guaranteed_profit = guaranteed_return - used_bankroll
        guaranteed_profit_pct = (guaranteed_profit / used_bankroll) * 100 if used_bankroll > 0 else 0.0
        return StakePlan(
            requested_bankroll=bankroll,
            bankroll=used_bankroll,
            total_implied_probability=total_implied,
            allocations=allocations,
            guaranteed_return=guaranteed_return,
            guaranteed_profit=guaranteed_profit,
            guaranteed_profit_pct=guaranteed_profit_pct,
            constraints_applied=constraints_applied,
            constraints_note="; ".join(notes),
        )
